### PR TITLE
Support pushing/pulling prompts

### PR DESCRIPTION
**Pull request by ankrgyl**
As a follow up to #362, this change adds support for pushing prompt definitions (in addition to tools) and pulling them, so you can save them in your repo.

### CODE REVIEW

File: js/src/functions/upload.ts [hunk 1]
```diff
@@ -374,18 +445,44 @@ function formatNameAndSlug(pieces: string[]) {
...
+  }
+
+  async getName(projectId: string): Promise<string> {
+    if (!(projectId in this.idToName)) {
+      const response = await _internalGetGlobalState()
+        .appConn()
+        .post_json("api/project/get", {
+          id: projectId,
+        });
+      const result = z.array(projectSchema).parse(response);
```
Comment (manugoyal) [inline js/src/functions/upload.ts:481]: You can use the `.nonempty()` constraint to make zod check that the array is nonempty too.

File: js/src/cli-util/pull.ts [hunk 2]
```diff
@@ -0,0 +1,303 @@
...
+    ),
+  );
+
+  for (const projectName of Object.keys(projectNameToFunctions)) {
+    const projectFile = path.join(
+      outputDir,
+      `${slugify(projectName, { lower: true, strict: true, trim: true })}.ts`,
+    );
+    const resolvedProjectFile = path.resolve(projectFile);
+    const fileExists = fs.existsSync(projectFile);
```
Comment (manugoyal) [inline js/src/cli-util/pull.ts:87]: Why use the synchronous fs api in an async function?

File: js/src/cli-util/pull.ts [hunk 3]
```diff
@@ -0,0 +1,303 @@
...
+      }
+    }
+
+    const projectFileContents = await makeProjectFile({
+      projectName,
+      fileName: projectFile,
+      functions: projectNameToFunctions[projectName],
+      hasSpecifiedFunction: !!args.slug || !!args.id,
+    });
+    fs.writeFileSync(projectFile, projectFileContents);
```
Comment (manugoyal) [inline js/src/cli-util/pull.ts:126]: same here

**Review by manugoyal — APPROVED** [review]

File: js/src/functions/upload.ts (hunk 1 above)
Comment (ankrgyl) [reply js/src/functions/upload.ts:481]: done

File: js/src/cli-util/pull.ts (hunk 2 above)
Comment (ankrgyl) [reply js/src/cli-util/pull.ts:88]: sure fixed

File: js/src/cli-util/pull.ts (hunk 3 above)
Comment (ankrgyl) [reply js/src/cli-util/pull.ts:127]: fixed

**Pull request merged**